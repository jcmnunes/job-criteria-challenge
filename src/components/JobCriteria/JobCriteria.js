import React, { Component } from 'react';
import remove from 'lodash.remove';
import Form from './Form';
import Summary from './Summary';
import Card from '../UI/Card';

class JobCriteria extends Component {
  state = {
    experience: '',
    educations: [],
    hours: {
      min: '',
      max: '',
    },
    isFormSaved: false,
  };

  handleInputChange = e => {
    const { educations } = this.state;
    const { target } = e;
    const { name } = target;
    const value = target.type === 'checkbox' ? target.checked : target.value;

    if (target.type === 'checkbox') {
      if (value) {
        this.setState({ educations: [...educations, name] });
      } else {
        const newEducations = [...educations];
        remove(newEducations, item => item === name);
        this.setState({ educations: newEducations });
      }
    } else if (name === 'experience') {
      this.setState({ experience: value });
    } else {
      this.setState({
        hours: {
          ...this.state.hours,
          [name]: value,
        },
      });
    }
  };

  handleSubmitSuccess = () => {
    this.setState({ isFormSaved: true });
  };

  handleEdit = () => {
    this.setState({ isFormSaved: false });
  };

  render() {
    const { isFormSaved } = this.state;
    return (
      <Card>
        {isFormSaved ? (
          <Summary handleEdit={this.handleEdit} formState={this.state} />
        ) : (
          <Form
            formState={this.state}
            handleSubmitSuccess={this.handleSubmitSuccess}
            onChange={this.handleInputChange}
          />
        )}
      </Card>
    );
  }
}

export default JobCriteria;
