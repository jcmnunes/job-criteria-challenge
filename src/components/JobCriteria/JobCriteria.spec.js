import React from 'react';
import { shallow } from 'enzyme';

import JobCriteria from './JobCriteria';

describe('Components', () => {
  describe('Form', () => {
    it('should render as expected, without crashing', () => {
      const wrapper = shallow(<JobCriteria />);

      expect(wrapper).toMatchSnapshot();
    });

    it('should render the Form is form is not saved', () => {
      const wrapper = shallow(<JobCriteria />);

      expect(wrapper.find('Form')).toHaveLength(1);
    });

    it('should render the Summary component when successfully submitting the form', () => {
      const wrapper = shallow(<JobCriteria />);
      wrapper.instance().handleSubmitSuccess();

      expect(wrapper.find('Summary')).toHaveLength(1);
    });

    it('should render the Form again if user wants to edit it', () => {
      const wrapper = shallow(<JobCriteria />);
      wrapper.instance().handleSubmitSuccess();
      expect(wrapper.find('Summary')).toHaveLength(1);

      wrapper.instance().handleEdit();
      expect(wrapper.find('Form')).toHaveLength(1);
    });

    it('should save the user input on state (experience)', () => {
      const wrapper = shallow(<JobCriteria />);
      const event = {
        target: {
          type: 'number',
          value: '4',
          name: 'experience',
        },
      };

      wrapper.instance().handleInputChange(event);

      expect(wrapper.state().experience).toBe('4');
    });

    it('should save the user input on state (educations)', () => {
      const wrapper = shallow(<JobCriteria />);
      const event = {
        target: {
          type: 'checkbox',
          checked: true,
          name: 'bachelor',
        },
      };
      wrapper.instance().handleInputChange(event);
      expect(wrapper.state().educations).toEqual(['bachelor']);

      const newEvent = {
        ...event,
        target: {
          ...event.target,
          checked: false,
        },
      };

      wrapper.instance().handleInputChange(newEvent);
      expect(wrapper.state().educations).toEqual([]);
    });

    it('should save the user input on state (hours)', () => {
      const wrapper = shallow(<JobCriteria />);
      const event = {
        target: {
          type: 'number',
          value: '3',
          name: 'max',
        },
      };

      wrapper.instance().handleInputChange(event);

      expect(wrapper.state().hours).toEqual({
        min: '',
        max: '3',
      });
    });
  });
});
