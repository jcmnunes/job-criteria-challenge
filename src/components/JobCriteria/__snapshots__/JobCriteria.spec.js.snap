// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components Form should render as expected, without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <JobCriteria />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Form
        formState={
          Object {
            "educations": Array [],
            "experience": "",
            "hours": Object {
              "max": "",
              "min": "",
            },
            "isFormSaved": false,
          }
        }
        handleSubmitSuccess={[Function]}
        onChange={[Function]}
      />,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "formState": Object {
          "educations": Array [],
          "experience": "",
          "hours": Object {
            "max": "",
            "min": "",
          },
          "isFormSaved": false,
        },
        "handleSubmitSuccess": [Function],
        "onChange": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-fjdhpX",
        "isStatic": true,
        "rules": Array [
          "
  background: white;
  box-shadow: 0 1px 3px hsla(0, 0%, 0%, 0.12), 0 1px 2px hsla(0, 0%, 0%, 0.24);
  border-radius: 4px;
  width: 640px;
  padding: 32px;
",
        ],
      },
      "displayName": "styled.div",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-fjdhpX",
      "target": "div",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Form
          formState={
            Object {
              "educations": Array [],
              "experience": "",
              "hours": Object {
                "max": "",
                "min": "",
              },
              "isFormSaved": false,
            }
          }
          handleSubmitSuccess={[Function]}
          onChange={[Function]}
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "formState": Object {
            "educations": Array [],
            "experience": "",
            "hours": Object {
              "max": "",
              "min": "",
            },
            "isFormSaved": false,
          },
          "handleSubmitSuccess": [Function],
          "onChange": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-fjdhpX",
          "isStatic": true,
          "rules": Array [
            "
  background: white;
  box-shadow: 0 1px 3px hsla(0, 0%, 0%, 0.12), 0 1px 2px hsla(0, 0%, 0%, 0.24);
  border-radius: 4px;
  width: 640px;
  padding: 32px;
",
          ],
        },
        "displayName": "styled.div",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-fjdhpX",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
